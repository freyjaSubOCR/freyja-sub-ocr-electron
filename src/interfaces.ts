interface RectPos {
    top: number
    left: number
    bottom: number
    right: number
}

interface RenderedVideo {
    timestamp: number
    keyFrame: boolean
    data: Buffer | Array<Buffer>
}

class ASSStyle {
    name = 'Default'
    fontname = '方正准圆_GBK'
    fontsize = '75'
    primaryColour = '&H00FFFFFF'
    secondaryColour = '&HF0000000'
    outlineColour = '&H00193768'
    backColour = '&HF0000000'
    bold = false
    italic = false
    underline = false
    strikeOut = false
    scaleX = 100
    scaleY = 100
    spacing = 0
    angle = 0
    borderStyle = 1
    outline = 2
    shadow = 0
    alignment = 2
    marginL = 10
    marginR = 10
    marginV = 15
    encoding = 1
}

export { RectPos, RenderedVideo, ASSStyle }
