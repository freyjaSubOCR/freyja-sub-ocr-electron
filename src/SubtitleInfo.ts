import { isNumber, toInteger } from 'lodash'
import { v4 as uuidv4 } from 'uuid'

// eslint-disable-next-line @typescript-eslint/interface-name-prefix
interface ISubtitleInfo {
    startFrame: number;
    endFrame: number;
    texts?: string[];
    startTime?: string;
    endTime?: string;
    box?: Int32Array;
}

class SubtitleInfo implements ISubtitleInfo {
    startFrame: number
    endFrame: number
    texts: string[] = []
    startTime?: string
    endTime?: string
    box?: Int32Array
    id: string
    fps?: number

    constructor(subtitleInfo: ISubtitleInfo)

    constructor(startFrame: number, endFrame?: number)

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    constructor(startFrame: any, endFrame?: number) {
        this.id = uuidv4()
        if (isNumber(startFrame)) {
            if (startFrame === undefined || endFrame === undefined) {
                throw new Error('Cannot init class from the provided parameters')
            }
            this.startFrame = startFrame
            this.endFrame = endFrame
        } else {
            const subtitleInfo = startFrame as ISubtitleInfo
            this.startFrame = subtitleInfo.startFrame
            this.endFrame = subtitleInfo.endFrame
            this.texts = subtitleInfo.texts === undefined ? [] : subtitleInfo.texts
            this.startTime = subtitleInfo.startTime
            this.endTime = subtitleInfo.endTime
            this.box = subtitleInfo.box
        }
    }

    get text(): string | undefined {
        const freq = {} as Record<string, number>
        let maxFreq = 0
        let maxFreqText = ''
        if (this.texts === undefined || this.texts.length === 0) {
            return undefined
        }
        for (const text of this.texts) {
            freq[text] = text in freq ? freq[text] + 1 : 1
            if (freq[text] > maxFreq) {
                maxFreq = freq[text]
                maxFreqText = text
            }
        }
        return maxFreqText
    }

    set text(value) {
        if (value !== undefined) {
            this.texts = [value]
        }
    }

    GenerateTime(fps: number) {
        this.fps = fps

        let timeInt = Math.floor(this.startFrame * 1000 / fps)
        let timeStruct = new Date(timeInt)
        this.startTime = `${timeStruct.getUTCHours().toString().padStart(2, '0')}:${timeStruct.getUTCMinutes().toString().padStart(2, '0')}:${timeStruct.getUTCSeconds().toString().padStart(2, '0')}.${Math.floor(timeStruct.getUTCMilliseconds() / 10).toString().padStart(2, '0')}`

        timeInt = Math.floor(this.endFrame * 1000 / fps)
        timeStruct = new Date(timeInt)
        this.endTime = `${timeStruct.getUTCHours().toString().padStart(2, '0')}:${timeStruct.getUTCMinutes().toString().padStart(2, '0')}:${timeStruct.getUTCSeconds().toString().padStart(2, '0')}.${Math.floor(timeStruct.getUTCMilliseconds() / 10).toString().padStart(2, '0')}`
    }

    get startTimeValidated() {
        if (this.fps !== undefined) {
            this.GenerateTime(this.fps)
        }
        return this.startTime
    }

    set startTimeValidated(value) {
        if (value === undefined) {
            this.startTime = undefined
        } else {
            const match = value.match(/^(\d{2}):([0-5]\d):([0-5]\d).(\d{2})$/)
            if (match) {
                this.startTime = value
                if (this.fps !== undefined) {
                    this.startFrame = Math.round(((toInteger(match[1]) * 60 + toInteger(match[2])) * 60 + toInteger(match[3]) + toInteger(match[4]) / 100) * this.fps)
                }
            }
        }
    }

    get endTimeValidated() {
        if (this.fps !== undefined) {
            this.GenerateTime(this.fps)
        }
        return this.endTime
    }

    set endTimeValidated(value) {
        if (value === undefined) {
            this.endTime = undefined
        } else {
            const match = value.match(/^(\d{2}):([0-5]\d):([0-5]\d).(\d{2})$/)
            if (match) {
                this.endTime = value
                if (this.fps !== undefined) {
                    this.endFrame = Math.round(((toInteger(match[1]) * 60 + toInteger(match[2])) * 60 + toInteger(match[3]) + toInteger(match[4]) / 100) * this.fps)
                }
            }
        }
    }
}

export { ISubtitleInfo, SubtitleInfo }
