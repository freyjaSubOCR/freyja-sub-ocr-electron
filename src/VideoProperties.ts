import { isNumber } from 'lodash'

// eslint-disable-next-line @typescript-eslint/interface-name-prefix
interface IVideoProperties {
    duration: number;
    timeBase: number[];
    fps: number[];
    width: number;
    height: number;
}

class VideoProperties implements IVideoProperties {
    duration: number
    timeBase: number[]
    fps: number[]
    width: number
    height: number

    constructor(videoProperties: IVideoProperties)

    constructor(duration: number, timeBase?: number[], fps?: number[], width?: number, height?: number)

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    constructor(duration: any, timeBase?: number[], fps?: number[], width?: number, height?: number) {
        if (isNumber(duration)) {
            if (duration === undefined || timeBase === undefined || fps === undefined || width === undefined || height === undefined) {
                throw new Error('Cannot init class from the provided parameters')
            }
            this.duration = duration
            this.timeBase = timeBase
            this.fps = fps
            this.width = width
            this.height = height
        } else {
            const videoProperties = duration as IVideoProperties
            this.duration = videoProperties.duration
            this.timeBase = videoProperties.timeBase
            this.fps = videoProperties.fps
            this.width = videoProperties.width
            this.height = videoProperties.height
        }
    }

    get unitFrame(): number {
        return this.timeBase[1] *
                this.fps[1] /
                this.timeBase[0] /
                this.fps[0]
    }

    get lastFrame(): number {
        return Math.floor(this.duration / this.unitFrame) - 1
    }
}

export { IVideoProperties, VideoProperties }
