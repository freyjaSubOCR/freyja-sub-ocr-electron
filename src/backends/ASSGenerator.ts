import { ipcMain } from 'electron'
import { ASSStyle } from '@/interfaces'
import { SubtitleInfo } from '@/SubtitleInfo'
import { VideoProperties } from '@/VideoProperties'
import logger from '@/logger'

class ASSGenerator {
    private style = new ASSStyle()

    registerIPCListener(): void {
        ipcMain.handle('ASSGenerator:Generate', async (e, ...args) => {
            try {
                return this.Generate(args[0], args[1])
            } catch (error) {
                logger.error(error.message)
                return null
            }
        })
    }

    Generate(subtitleInfos: SubtitleInfo[], videoProperties: VideoProperties): string {
        const strings: string[] = []
        strings.push('[Script Info]')
        strings.push('ScriptType: v4.00+')
        strings.push(`PlayResX: ${videoProperties.width}`)
        strings.push(`PlayResY: ${videoProperties.height}`)
        strings.push('[V4+ Styles]')
        strings.push('Format: Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, Alignment, MarginL, MarginR, MarginV, Encoding')
        strings.push(`Style: ${this.style.Name},${this.style.Fontname},${this.style.Fontsize},${this.style.PrimaryColour},${this.style.SecondaryColour},${this.style.OutlineColour},${this.style.BackColour},${this.style.Bold ? -1 : 0},${this.style.Italic ? -1 : 0},${this.style.Underline ? -1 : 0},${this.style.StrikeOut ? -1 : 0},${this.style.ScaleX},${this.style.ScaleY},${this.style.Spacing},${this.style.Angle},${this.style.BorderStyle},${this.style.Outline},${this.style.Shadow},${this.style.Alignment},${this.style.MarginL},${this.style.MarginR},${this.style.MarginV},${this.style.Encoding}`)

        strings.push('[Events]')
        strings.push('Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text')
        for (const subtitleInfo of subtitleInfos) {
            strings.push(`Dialogue: 0,${subtitleInfo.startTime},${subtitleInfo.endTime},Default,,0,0,0,,${subtitleInfo.text}`)
        }
        return strings.join('\n')
    }

    ApplyStyle(style: ASSStyle) {
        this.style = style
    }
}

export default ASSGenerator
