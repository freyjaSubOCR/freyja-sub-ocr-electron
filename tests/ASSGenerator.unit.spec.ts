import ASSGenerator from '@/backends/ASSGenerator'
import { ASSStyle } from '@/interfaces'
import { ISubtitleInfo, SubtitleInfo } from '@/SubtitleInfo'
import { VideoProperties } from '@/VideoProperties'
import fs_ from 'fs'
const fs = fs_.promises

describe('ASSGenerator.ts', () => {
    it('generate test', async () => {
        const ass = new ASSGenerator()
        ass.ApplyStyle(new ASSStyle())
        const subtitleInfos: SubtitleInfo[] = JSON.parse(await fs.readFile('tests/files/subtitleInfos.json', { encoding: 'utf-8' }))
            .map((element: ISubtitleInfo) => new SubtitleInfo(element))
        const assContent = ass.Generate(subtitleInfos, new VideoProperties(0, [], [], 1920, 1080))
        expect(assContent).toMatchSnapshot()
    }, 100000)
})
