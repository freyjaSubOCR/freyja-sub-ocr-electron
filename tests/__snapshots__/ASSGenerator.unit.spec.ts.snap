// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ASSGenerator.ts generate test 1`] = `
"[Script Info]
ScriptType: v4.00+
PlayResX: 1920
PlayResY: 1080
[V4+ Styles]
Format: Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, Alignment, MarginL, MarginR, MarginV, Encoding
Style: Default,方正准圆_GBK,75,&H00FFFFFF,&HF0000000,&H00193768,&HF0000000,0,0,0,0,100,100,0,0,1,2,0,2,10,10,15,1
[Events]
Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text
Dialogue: 0,00:00:00.08,00:00:01.12,Default,,0,0,0,,我是神野惠
Dialogue: 0,00:00:01.45,00:00:03.42,Default,,0,0,0,,是和泉纱雾的同班同学
Dialogue: 0,00:00:03.50,00:00:05.92,Default,,0,0,0,,你为什么要让她进我们家啊
Dialogue: 0,00:00:06.00,00:00:10.13,Default,,0,0,0,,跟我差不多岁数的女孩子都最喜欢鸡鸡了
Dialogue: 0,00:00:10.26,00:00:11.59,Default,,0,0,0,,最…最喜欢？
Dialogue: 0,00:00:11.84,00:00:15.51,Default,,0,0,0,,名字就叫「把小和泉从房间里拖出来！」同盟
Dialogue: 0,00:00:15.72,00:00:16.80,Default,,0,0,0,,不出去 不出去
Dialogue: 0,00:00:16.80,00:00:18.26,Default,,0,0,0,,我说不出去就是不出去
Dialogue: 0,00:00:18.39,00:00:21.39,Default,,0,0,0,,哥哥想让妹妹变成什么样？
Dialogue: 0,00:00:22.14,00:00:23.39,Default,,0,0,0,,和泉政宗？
Dialogue: 0,00:00:23.81,00:00:25.81,Default,,0,0,0,,居然是如此年轻的女孩子？
Dialogue: 0,00:00:26.19,00:00:29.98,Default,,0,0,0,,我的下一个作品要请情色漫画老师来担任插画师"
`;
